[LV] Support binary and unary operations with EVL-vectorization

nikolaypanchenko opened the pull request:
The patch adds `VPWidenEVLRecipe` which represents `VPWidenRecipe` + EVL argument. The new recipe replaces `VPWidenRecipe` in `tryAddExplicitVectorLength` for each binary and unary operations. Follow up patches will extend support for remaining cases, like `FCmp` and `ICmp`

nikolaypanchenko committed:
[LV] Support binary and unary operations with EVL-vectorization

The patch adds `VPWidenEVLRecipe` which represents `VPWidenRecipe` + EVL
argument. The new recipe replaces `VPWidenRecipe` in
`tryAddExplicitVectorLength` for each binary and unary operations.
Follow up patches will extend support for remaining cases, like `FCmp`
and `ICmp`

npanchen commented:
CC: @ppenzin

arcbbb commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -993,6 +997,64 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+  assert(is_contained(operands(), Op) && "Op must be an operand of the recipe");
+  // EVL in that recipe is always the last operand, thus any use before means
+  // the VPValue should be vectorized.
+  for (unsigned I = 0, E = getNumOperands() - 1; I != E; ++I)
+    if (getOperand(I) == Op)
+      return false;
+  return true;
```
can this be simplified to `return getOperand(getNumOperands() - 1) == Op;` ?

nikolaypanchenko replied:
It's _technically_ possible to use `EVL` as first or second operand in the operation itself:
```
%evl = ...
%0 = add %evl, %0
```
i.e. that check makes sure `onlyFirstLaneUsed` returns false if `Op == EVL` for the example above

arcbbb replied:
yes, looks like this corresponds to fhahn's comment about if (!State.get(Op0, 0)->getType()->isVectorTy()), regarding whether operands can be scalar."

fhahn replied:
If EVL would be the first/second operand, it would implicitly be splatted. In any case, no such plans can be constructed at the moment I think. Better to verify EVL is only used in supported places?

nikolaypanchenko replied:
I will add verification

arcbbb commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1363,6 +1370,49 @@ class VPWidenRecipe : public VPRecipeWithIRFlags {
+
+  VPWidenRecipe *clone() override final {
+    SmallVector<VPValue *> Ops(operands());
+    VPValue *EVL = Ops.pop_back_val();
```
Use `getEVL()` directly

nikolaypanchenko replied:
The intention is to remove `EVL` operand from `Ops` to be able to reuse `VPWidenEVLRecipe` ctor

arcbbb commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1296,6 +1297,56 @@ void VPlanTransforms::addActiveLaneMask(
+static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+  DenseSet<VPRecipeBase *> ToRemove;
+
+  ReversePostOrderTraversal<VPBlockDeepTraversalWrapper<VPBlockBase *>> RPOT(
+      Plan.getEntry());
```
Consider utilizing `Plan.getVectorLoopRegion()->getEntryBasicBlock()` to replace recipes that are within the loop.

nikolaypanchenko replied:
That's a good point. We really need to traverse entire VPlan, but replace instructions in postexit with init vl, like for a reduction.
That said, I need to add special check that EVL is reachable before replacing.

arcbbb replied:
You're right; it reminds me that we must explicitly express InitVL in VPlan once we support reduction. This is because EVL from the last two iterations cannot be used in the post-exit for correctness. For this PR, can we temporarily skip the preheader and post-exit?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1333,13 +1335,18 @@ class VPInstruction : public VPRecipeWithIRFlags {
 /// where each ingredient transforms into a vectorized version of itself.
 class VPWidenRecipe : public VPRecipeWithIRFlags {
+protected:
   unsigned Opcode;
```
sufficient to leave this private and use `getOpcode` in the subclass?

fhahn reviewed (commented):
Just to double check, this isn't needed for correctness, right?

nikolaypanchenko commented:
> Just to double check, this isn't needed for correctness, right?

that's correct.

fhahn reviewed (commented):
> > Just to double check, this isn't needed for correctness, right?
> 
> that's correct.

Great, could you add more detail about the motivation of the change to the PR description, i.e. why it is preferable to use the EVL version instead of regular wide ops?

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1363,6 +1370,49 @@ class VPWidenRecipe : public VPRecipeWithIRFlags {
+  const VPValue *getEVL() const { return getOperand(getNumOperands() - 1); }
+
+  /// A helper function to create widen EVL recipe from regular widen recipe.
+  static VPWidenEVLRecipe *create(VPWidenRecipe *W, VPValue &EVL);
```
Not sure I understand. There is `VPWidenEVLRecipe(Instruction &I, iterator_range<IterT> Operands, VPValue &EVL)` ctor defined above

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -993,6 +997,64 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+  Value *VPInst = nullptr;
+
+  //===------------------- Binary and Unary Ops ---------------------===//
+  if (Instruction::isBinaryOp(Opcode) || Instruction::isUnaryOp(Opcode)) {
```
there's `llvm_unreachable` at 1042. Later I plan to add support for remaining opcodes: FCmp and ICmp here.

fhahn replied:
👍🏻

nikolaypanchenko committed:
Addressed comments

nikolaypanchenko committed:
Rebase

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -35,6 +36,11 @@ class VPlanVerifier {
+  // Verify that \p EVL is used correctly. The user must be  either in EVL-based
+  // recipes as a last operand or VPInstruction::Add which is incoming value
+  // into EVL's recipe.
+  bool verifyEVLRecipe(const VPInstruction &EVL) const;
```
Could this be split off, and submitted first?

nikolaypanchenko committed:
Removed evl verification

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1366,6 +1373,49 @@ class VPWidenRecipe : public VPRecipeWithIRFlags {
+
+  ~VPWidenEVLRecipe() override = default;
+
+  VPWidenRecipe *clone() override final {
```
at the moment this is only used when vectorizing the epilogue, which is disabled for EVL-style vectorization for now, right? Other recipes that don't exercise the clone() use unreachable here.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8457,14 +8457,15 @@ void LoopVectorizationPlanner::buildVPlansWithVPRecipes(ElementCount MinVF,
       VPlanTransforms::optimize(*Plan, *PSE.getSE());
       // TODO: try to put it close to addActiveLaneMask().
       // Discard the plan if it is not EVL-compatible
-      if (CM.foldTailWithEVL() &&
```
This is independent of the unary/binary support, right? Can be split off + a test that forces VF =1 and IC> 1?

nikolaypanchenko replied:
It does relate to that change. Without it we create EVL recipes for scalar VPlan. For load/store that is not a problem, but it's a problem for `WidenEVL` since it cannot handle VF=1

fhahn replied:
Hm, does it make sense to create VPlans with EVL in general? IIUC it doesn't and this could be done independent of the patch.

widen loads and stores shouldn't be created in a scalar plan, I guess at the moment this just works as there are no recipes that can be in scalar plans and get converted to an EVL version without this patch.

nikolaypanchenko replied:
> Hm, does it make sense to create VPlans with EVL in general? IIUC it doesn't and this could be done independent of the patch.

Not sure I follow. 

> widen loads and stores shouldn't be created in a scalar plan, I guess at the moment this just works as there are no recipes that can be in scalar plans and get converted to an EVL version without this patch.

I double checked the code and looks like I got confused with `VPInstruction`. So I agree `VPWidenRecipe` won't be a problem. Will move it to a separate patch

fhahn replied:
Sounds great!

LiqinWeng replied:
> IIU
what is IIUC ？

nikolaypanchenko committed:
Moved clone() methods to unreachable for EVL-recipes

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1385,6 +1387,57 @@ void VPlanTransforms::addActiveLaneMask(
+static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+  VPDominatorTree VPDT;
+  VPDT.recalculate(Plan);
+  DenseSet<VPRecipeBase *> ToRemove;
```
SmallVector should be sufficient?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1385,6 +1387,57 @@ void VPlanTransforms::addActiveLaneMask(
+  for (VPBasicBlock *VPBB :
+       reverse(VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT))) {
+    for (VPRecipeBase &R : make_early_inc_range(reverse(*VPBB))) {
+      if (!properlyDominates(EVL.getDefiningRecipe(), &R, VPDT))
```
Can this be an assert? We just earlier inserted EVL at a place that should dominate all recipes handled below?

nikolaypanchenko replied:
It's related to [comment above](https://github.com/llvm/llvm-project/pull/93854#discussion_r1626100405): there are some cases when preheader or postexit may technically contain evl-based recipes, but they do require EVL, but it should be constructed before.

nikolaypanchenko replied:
will remove it for now as it does assert work for VF=1.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1385,6 +1387,57 @@ void VPlanTransforms::addActiveLaneMask(
+      Plan.getEntry());
+  DenseSet<VPValue *> HeaderMasks = collectAllHeaderMasks(Plan);
+  for (VPBasicBlock *VPBB :
+       reverse(VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT))) {
```
Is an RPOT & reverse needed?

nikolaypanchenko replied:
Yes, otherwise recipes won't be removed correctly.
BTW, the comment I removed was copied with a loop from `removeDeadRecipes`

fhahn replied:
I am not sure, I think the reason removeDeadRecipes does it in that order to remove later dead recipes before their users, but the current patch removes all replaced recipes, then removes the masks. The only operand that isn't used any longer when removing the original recipes is the mask, so I am not sure in what cases the order matters?

nikolaypanchenko replied:
agree. removed

nikolaypanchenko committed:
Addressed comments

nikolaypanchenko committed:
Addressed comments

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1375,6 +1382,45 @@ class VPWidenRecipe : public VPRecipeWithIRFlags {
+  VPValue *getEVL() { return getOperand(getNumOperands() - 1); }
+  const VPValue *getEVL() const { return getOperand(getNumOperands() - 1); }
+
+  /// Produce widened copies of all Ingredients.
```
nit: there is only a single ingredient

fhahn replied:
I realized that the `VPWidenRecipe` comment was also out-of-date. Updated in e6fdecd29059c0951a4970e9e6cf478972844099, would probably be good to update the comment here as well

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1051,6 +1055,55 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+                          "explicit vector length.");
+  VPValue *Op0 = getOperand(0);
+
+  // If it's scalar operation, hand translation over to VPWidenRecipe
```
Comment out of sync?

fhahn replied:
Comment still needs to be dropped?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1051,6 +1055,55 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+                                             "vp.op");
+
+    if (I)
+      if (auto *VecOp = dyn_cast<Instruction>(VPInst))
```
nit: can dyn_cast_or_null be used instead of `if (I)`?

fhahn replied:
Still pending?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1402,6 +1403,53 @@ void VPlanTransforms::addActiveLaneMask(
+  ReversePostOrderTraversal<VPBlockDeepTraversalWrapper<VPBlockBase *>> RPOT(
+      Plan.getEntry());
+  DenseSet<VPValue *> HeaderMasks = collectAllHeaderMasks(Plan);
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT)) {
```
RPOT still needed?

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-intrinsics.ll:
```diff
@@ -31,7 +31,7 @@ define void @foo(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %N) {
 ; IF-EVL-NEXT:    vp<[[PTR2:%[0-9]+]]> = vector-pointer ir<[[GEP2]]>
 ; IF-EVL-NEXT:    WIDEN ir<[[LD2:%.+]]> = vp.load vp<[[PTR2]]>, vp<[[EVL]]>
-; IF-EVL-NEXT:    WIDEN ir<[[ADD:%.+]]> = add nsw ir<[[LD2]]>, ir<[[LD1]]>
+; IF-EVL-NEXT:    WIDEN-VP ir<[[ADD:%.+]]> = add nsw ir<[[LD2]]>, ir<[[LD1]]>
```
Should this also match the EVL operand?

nikolaypanchenko committed:
Don't use RPOT + rebase

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1405,11 +1407,16 @@ class VPInstruction : public VPRecipeWithIRFlags {
   VPWidenRecipe(Instruction &I, iterator_range<IterT> Operands)
-      : VPRecipeWithIRFlags(VPDef::VPWidenSC, Operands, I),
-        Opcode(I.getOpcode()) {}
+      : VPWidenRecipe(VPDef::VPWidenSC, I, Operands) {}
```
This should probably have a custom `classof` is `isa<VPWidenRecipe>` returns true for both VPWidenRecipe and VPWidenEVLRecipe?

nikolaypanchenko replied:
I assume than you really meant to introduce base class for `VPWidenRecipe` and `VPWidenEVLRecipe` that will return true for both of them, right ? In this case class hierarchy will be similar to `VPWiden[Load|Store][|EVL]`.
If so, same should go to all future EVL-recipes:
```
           VPSomeRecipeBase
           /             \
VPSomeRecipe          VPSomeEVLRecipe
```

```
VPSomeRecipeBase::classof(...) { return Opcode == VPSomeRecipeSC || Opcode == VPSomeEVLRecipeSC; }
```

fhahn replied:
Ideally VPWidenRecipe could serve as such a base class as mentioned above, unless there’s a compelling reason not to. That way we automatically benefit from all folds and analysis already implemented for VPWidenRecipe

nikolaypanchenko replied:
Not sure I understand added hierarchy then. If the goal is to allow reuse of existing code, then what is different in `VPWidenStoreEVL` or `VPWidenLoadEVL` as they won't be treated as `VPWidenStore` or `VPWidenLoad` respectively: 
```
isa<VPWidenMemoryRecipe>(EVLStore) => true
isa<VPWidenStoreEVLRecipe>(EVLStore) => true
isa<VPWidenStoreRecipe>(EVLStore) => false
```
Anyway, I'm _ok_ to extend classof for now

fhahn replied:
Thanks for the update. Adjusting `VPWidenStoreEVLRecipe`/`VPWidenLoadEVLRecipe` would probably make sense separately, although the reasoning may be a bit more complicated.

LiqinWeng commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1140,6 +1144,55 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+
+    SmallVector<Value *, 4> Ops;
+    for (unsigned I = 0, E = getNumOperands() - 1; I < E; ++I) {
+      VPValue *VPOp = getOperand(I);
```
```
for (VPValue *VPOp : operands()) {
      bool IsScalar = VPOp == EVL;
      Ops.push_back(State.get(VPOp, 0, IsScalar));
 }
```

nikolaypanchenko committed:
Addressed comments + rebase

LiqinWeng commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1405,6 +1407,66 @@ void VPlanTransforms::addActiveLaneMask(
+static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+  VPDominatorTree VPDT;
+  VPDT.recalculate(Plan);
+  SmallVector<VPRecipeBase *> ToRemove;
```
Unused， pls remove

LiqinWeng commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1140,6 +1144,53 @@ void VPWidenRecipe::execute(VPTransformState &State) {
 }
 
+void VPWidenEVLRecipe::execute(VPTransformState &State) {
+  assert(State.UF == 1 && "Expected only UF == 1 when vectorizing with "
```
Sorry, I don't know much about vectorized design. I would like to know why UF is forced to be set to 1?

nikolaypanchenko replied:
It's only for EVL-vectorization, as for now it's easier not to handle multiple `get.vector.length`, where one or many could return `0`, thus following instructions will process 0 elements. Another worry I have is we will have to emit extra checks for zero EVL if instruction does ignore it, like RISC-V's `vmv.x.s` if it's generated within a loop

LiqinWeng commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1140,6 +1144,53 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+  if (Instruction::isBinaryOp(Opcode) || Instruction::isUnaryOp(Opcode)) {
+    // Just widen unops and binops.
+
+    SmallVector<Value *, 4> Ops;
```
add: State.setDebugLocFrom(getDebugLoc())

LiqinWeng commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2642,6 +2694,11 @@ struct VPWidenStoreEVLRecipe final : public VPWidenMemoryRecipe {
   }
 
+  VPWidenStoreEVLRecipe *clone() override {
+    llvm_unreachable("VPWidenStoreEVLRecipe cannot be cloned");
```
I don't understand the purpose of this interface. I want to know why it is set to unreachable

nikolaypanchenko replied:
[Florian's comment on this function](https://github.com/llvm/llvm-project/pull/93854#discussion_r1643384493). If you do have a use of a clone function for other cases, it will be great to have it in upstream.

nikolaypanchenko committed:
Removed dead code, moved EVL to last args in EVL recipes, added debugloc

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2558,14 +2605,19 @@ struct VPWidenLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
 /// mask.
 struct VPWidenLoadEVLRecipe final : public VPWidenMemoryRecipe, public VPValue {
-  VPWidenLoadEVLRecipe(VPWidenLoadRecipe &L, VPValue &EVL, VPValue *Mask)
+  VPWidenLoadEVLRecipe(VPWidenLoadRecipe &L, VPValue *Mask, VPValue &EVL)
```
revert

nikolaypanchenko committed:
Addressed latest comments

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1140,6 +1144,54 @@ void VPWidenRecipe::execute(VPTransformState &State) {
+}
+
+bool VPWidenEVLRecipe::onlyFirstLaneUsed(const VPValue *Op) const {
+  assert(is_contained(operands(), Op) && "Op must be an operand of the recipe");
```
now that this is quite compact, maybe better to move to header again, consistent with other `onlyFirstLaneUsed` implementations?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1405,6 +1407,62 @@ void VPlanTransforms::addActiveLaneMask(
+              })
+              .Default([&](VPRecipeBase *R) { return nullptr; });
+
+      if (NewRecipe) {
```
nit
```suggestion
      if (!NewRecipe) 
        continue;
```
to reduce indent below

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reverse-load-store.ll:
```diff
@@ -16,46 +16,46 @@ define void @reverse_load_store(i64 %startval, ptr noalias %ptr, ptr noalias %pt
 ; IF-EVL-NEXT:    [[TMP1:%.*]] = mul i64 [[TMP0]], 4
-; IF-EVL-NEXT:    [[TMP4:%.*]] = sub i64 [[TMP1]], 1
-; IF-EVL-NEXT:    [[N_RND_UP:%.*]] = add i64 1024, [[TMP4]]
+; IF-EVL-NEXT:    [[TMP2:%.*]] = sub i64 [[TMP1]], 1
```
Are the changes here caused by the patch? After a quick spot check, it seems only related to CHECK variable renames? If so would be good to remove them from the patch

nikolaypanchenko replied:
most likely is an artifact from some of my previous commits

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1405,6 +1407,62 @@ void VPlanTransforms::addActiveLaneMask(
+                return new VPWidenEVLRecipe(W, EVL);
+              })
+              .Case<VPReductionRecipe>([&](VPReductionRecipe *Red) {
+                return new VPReductionEVLRecipe(*Red, EVL,
+                                                GetNewMask(Red->getCondOp()));
```
```suggestion
                VPValue *NewMask = GetNewMask(Red-> getCondOp());
                return new VPReductionEVLRecipe(*Red, EVL, NewMask);
```

to be consistent with other uses of GetNewMask

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2566,6 +2623,11 @@ struct VPWidenLoadEVLRecipe final : public VPWidenMemoryRecipe, public VPValue {
   }
 
+  VPWidenLoadEVLRecipe *clone() override {
+    llvm_unreachable("VPWidenLoadEVLRecipe cannot be cloned");
```
This and `VPWidenStoreEVLRecipe *clone() override` are unrelated to the patch IIUC, would be good to split off

fhahn reviewed (commented):
Thanks for the latest updates, a few final comments inline

nikolaypanchenko committed:
Addressed comments

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1443,6 +1458,54 @@ class VPWidenRecipe : public VPRecipeWithIRFlags {
+  bool onlyFirstLaneUsed(const VPValue *Op) const override {
+    assert(is_contained(operands(), Op) &&
+           "Op must be an operand of the recipe");
+    // EVL in that recipe is always the last operand, thus any use before means
```
Would be good to post a follow-up patch to enforce EVL only used as last operand of various recipes in the verifier

nikolaypanchenko replied:
Sure. I will create PR with verification I added and removed previously

fhahn replied:
Great thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1214,6 +1218,47 @@ InstructionCost VPWidenRecipe::computeCost(ElementCount VF,
+                                             "vp.op");
+
+    if (auto *VecOp = dyn_cast_or_null<Instruction>(VPInst))
+      VecOp->copyIRFlags(I);
```
This should use the flags of the recipe, which may need to be dropped for correctness

```suggestion
      setFlags(VecOp);
```

See `@drop_vector_nuw_nsw` in `llvm/test/Transforms/LoopVectorize/X86/drop-poison-generating-flags.ll`, might be good to have a similar test for EVL widen recipes

nikolaypanchenko replied:
Current test won't be vectorized with EVL since it uses widen induction. https://github.com/llvm/llvm-project/pull/82021 is needed first to unblock widen EVL induction.
That said, can this test be added later ?

fhahn replied:
Sure, test could also be added before widened EVL support.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1214,6 +1218,47 @@ InstructionCost VPWidenRecipe::computeCost(ElementCount VF,
+  VPValue *EVL = getEVL();
+  Value *EVLArg = State.get(EVL, 0, /*NeedsScalar=*/true);
+  unsigned Opcode = getOpcode();
+  Instruction *I = getUnderlyingInstr();
```
```suggestion
  Instruction *UI = getUnderlyingInstr();
```

to avoid clash with `I` below. 

The underlying instruction is only used by `State.addMetadata`, might be better to use`dyn_cast_or_null<Instruction>(getUnderlyingValue())` to make the underlying instruction optional. (getUnderlyingInstr` uses `cast<Instruction>`, which makes the underlying instruction non-optional)

nikolaypanchenko committed:
Refactored code and removed flag propagation as they're not yet
supported by vp intrinsics

nikolaypanchenko committed:
Only set FMFs on vp intrinsics

VP intrinsics can only accept FMFs at this moment, thus trying to set
other flags will lead to ICE

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1208,6 +1212,43 @@ InstructionCost VPWidenRecipe::computeCost(ElementCount VF,
+  Builder.setMask(Mask).setEVL(EVLArg);
+  Value *VPInst =
+      Builder.createVectorInstruction(Opcode, Ops[0]->getType(), Ops, "vp.op");
+  if (isa<FPMathOperator>(VPInst))
```
IIUC only FMFs can be set here as other flags are not supported on calls. Would be good to document here

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -262,7 +262,8 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
             return inferScalarType(R->getOperand(0));
           })
           .Case<VPBlendRecipe, VPInstruction, VPWidenRecipe, VPReplicateRecipe,
-                VPWidenCallRecipe, VPWidenMemoryRecipe, VPWidenSelectRecipe>(
+                VPWidenCallRecipe, VPWidenMemoryRecipe, VPWidenSelectRecipe,
+                VPWidenEVLRecipe>(
```
nit: keep VPWidenEVLRecipe next to VPWidenRecipe?

fhahn reviewed (approved):
LGTM, thanks! A few small additional suggestions inline

nikolaypanchenko committed:
Rebase

LiqinWeng commented:
@nikolaypanchenko can merge this patch？

kazutakahirata commented:
I've fixed an unused variable warning from this PR with ce192b87b2a09ee27e4077763db0486921a485c0.

nikolaypanchenko commented:
> I've fixed an unused variable warning from this PR with [ce192b8](https://github.com/llvm/llvm-project/commit/ce192b87b2a09ee27e4077763db0486921a485c0).

thanks. I'll double check why I missed it in the first place
